sympy/sets/tests/test_fancysets.py

```diff
@@ -140,20 +151,74 @@ def test_Range():
     assert Range(-oo, 0)
     assert Range(0, -oo, -1)
 
-    assert Range(0, oo, 2)._last_element is oo
-    assert Range(-oo, 1, 1)._last_element is S.Zero
+    assert Range(0, oo, 2)[-1] == oo
+    assert Range(-oo, 1, 1)[-1] is S.Zero
+    assert Range(oo, 1, -1)[-1] == 2
+    assert Range(0, -oo, -2)[-1] == -oo
+    assert Range(1, 10, 1)[-1] == 9
 
     it = iter(Range(-oo, 0, 2))
-    assert (next(it), next(it)) == (-2, -4)
+    raises(ValueError, lambda: next(it))
 
     assert Range(-1, 10, 1).intersect(S.Integers) == Range(-1, 10, 1)
     assert Range(-1, 10, 1).intersect(S.Naturals) == Range(1, 10, 1)
 
-    assert Range(1, 10, 1)._ith_element(5) == 6 # the index starts from zero
-    assert Range(1, 10, 1)._last_element == 9
+    # test _last_element via indexing
+    assert Range(1, 10, 1)[5] == 6
+    assert Range(1, 12, 2)[5] == 11
+    assert Range(1, 10, 1)[-1] == 9
+    assert Range(1, 10, 3)[-1] == 7
+    raises(ValueError, lambda: Range(1, oo)[-2])
+    raises(ValueError, lambda: Range(-oo, 1)[2])
+    raises(IndexError, lambda: Range(10)[-20])
+    raises(IndexError, lambda: Range(10)[20])
+
+    # test slicing
+    raises(ValueError, lambda: Range(2, -oo, -2)[2:2:0])
+    assert Range(2, -oo, -2)[2:2:2] is S.EmptySet
+    assert Range(2, -oo, -2)[:2:2] == Range(0, 4, 4)
```

Comment:
- [ ] check that `Range(2,-oo,-2)[:4:2] == Range(2, -2, -2)`; the way this test is written makes it look like th estep will be positive